Tella/Scenes/Settings/ViewModel/UwaziServerViewModel.swift

```diff
@@ -0,0 +1,351 @@
+                receiveValue: { result in
+                    self.isLoading = false
+                    if result.0.success {
+                        self.showNextLanguageSelectionView = true
+                        if let token = result.1?.value(forHTTPHeaderField: "Set-Cookie") {
+                            let filteredToken = token.split(separator: ";")
+                            let connectId = filteredToken.first!.replacingOccurrences(of: "connect.sid=", with: "")
+                            self.token = connectId
+                        }
+                    }
+                }
+            )
+            .store(in: &subscribers)
+    }
+
+    func twoFactorAuthentication() {
+
+        guard let baseURL = serverURL.getBaseURL() else { return }
+
+        isLoading = true
+
+        UwaziServerRepository().twoFactorAuthentication(username: username, password: password, token: code, serverURL: baseURL)
+            .receive(on: DispatchQueue.main)
+            .sink(
+                receiveCompletion: { completion in
+
+                    switch completion {
+                    case .failure(let error):
+                        switch error {
+
+                        case .invalidURL:
+                            self.shouldShowAuthenticationError = true
+                            self.codeErrorMessage = error.errorDescription ?? ""
+                            self.isLoading = false
+                        case .httpCode(let code):
+                            // if the status code is 401 then the 2FA code is incorrect
+                            if code == 401 {
+                                self.shouldShowAuthenticationError = true
+                                self.codeErrorMessage = "Two-factor authentication failed."
+                                self.isLoading = false
+                            } else {
+                                self.shouldShowAuthenticationError = true
+                                self.codeErrorMessage = error.errorDescription ?? ""
+                                self.isLoading = false
+                            }
+                        case .unexpectedResponse:
+                            self.shouldShowAuthenticationError = true
+                            self.codeErrorMessage = error.errorDescription ?? ""
+                            self.isLoading = false
+                        }
+
+
+                    case .finished:
+                        self.shouldShowAuthenticationError = false
+                        self.codeErrorMessage = ""
+                        self.isLoading = false
+                        break
+
+                    }
+                },
+                receiveValue: { result in
+                    self.isLoading = false
+                    if result.0.success {
+                        self.showNextLanguageSelectionView = true
+                        if let token = result.1?.value(forHTTPHeaderField: "Set-Cookie") {
```

Comment:
Could you centralise this code, I think it's better to have the response ready in the view model